src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Match.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+using Microsoft.PowerFx.Core.App.ErrorContainers;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+
+#pragma warning disable SA1402 // File may only contain a single type
+#pragma warning disable SA1649 // File name should match first type name
+
+namespace Microsoft.PowerFx.Core.Texl.Builtins
+{
+    // Match(text:s, regular_expression:s, [options:s])
+    internal class MatchFunction : BaseMatchFunction
+    {
+        public MatchFunction(RegexCache regexCache)
+            : base("Match", TexlStrings.AboutMatch, DType.EmptyRecord, regexCache)
+        {
+        }
+    }
+
+    // MatchAll(text:s, regular_expression:s, [options:s])
+    internal class MatchAllFunction : BaseMatchFunction
+    {
+        public MatchAllFunction(RegexCache regexCache)
+            : base("MatchAll", TexlStrings.AboutMatchAll, DType.EmptyTable, regexCache)
+        {
+        }
+    }
+
+    internal class BaseMatchFunction : BuiltinFunction
+    {        
+        private readonly ConcurrentDictionary<string, Tuple<DType, bool, bool, bool>> _regexTypeCache;
```

Comment:
>ConcurrentDictionary [](http://example.com/codeflow?start=25&length=20)

In a future PR, this could now just be RegexCache , and then direct dictionary manipulation (unsafe) could be moved to the class (Safer)

Reply:
Agreed on future PR